src/NuGet.Core/NuGet.Packaging/Signing/Authoring/SigningOptions.cs

```diff
@@ -33,28 +57,124 @@ public sealed class SigningOptions
         public ISignatureProvider SignatureProvider { get; }
 
         /// <summary>
-        /// Logger to be used to display the logs during the execution of signer actions.
+        /// Logger to be used to display the logs during the execution of signing actions.
         /// </summary>
         public ILogger Logger { get; }
 
-        public SigningOptions(string packageFilePath, string outputFilePath, bool overwrite, ISignatureProvider signatureProvider, ILogger logger)
+        /// <summary>Instantiates a new <see cref="SigningOptions" /> object.</summary>
+        /// <param name="inputPackageStream">A readable stream for the package that will be used as input for any
+        /// signing operation.</param>
+        /// <param name="outputPackageStream">A readable and writeable stream for the output package for any signing
+        /// operation.</param>
+        /// <param name="overwrite">A flag indicating if an existing signature should be overwritten.</param>
+        /// <param name="signatureProvider">A provider to create a Signature that can be added to the package.</param>
+        /// <param name="logger">A logger.</param>
+        /// <remarks>Signing operations cannot be done in place; therefore, <paramref name="inputPackageStream"/>
+        /// and <paramref name="outputPackageStream" /> should be different streams.</remarks>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="inputPackageStream" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="outputPackageStream" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="signatureProvider" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="logger" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="inputPackageStream" /> and
+        /// <paramref name="outputPackageStream"/> are the same object.</exception>
+        public SigningOptions(
+            Lazy<Stream> inputPackageStream,
+            Lazy<Stream> outputPackageStream,
+            bool overwrite,
+            ISignatureProvider signatureProvider,
+            ILogger logger)
         {
-            PackageFilePath = packageFilePath ?? throw new ArgumentNullException(nameof(packageFilePath));
-            OutputFilePath = outputFilePath ?? throw new ArgumentNullException(nameof(outputFilePath));
+            _inputPackageStream = inputPackageStream ?? throw new ArgumentNullException(nameof(inputPackageStream));
```

Comment:
is it possible to send the same stream as input an output? Is there a way we can avoid this?

Reply:
It's possible to pass the same stream for both input and output, but independently reading and writing to the same stream would be disastrous.  XML Doc Comments [say](https://github.com/NuGet/NuGet.Client/pull/2065/files/7e26c8d493c2499e3fabfca1564c1d57812c16e6#diff-8618b88e754b06293304d2f1f7dd5924R72) the streams should be different.

We could perform a reference equality check on the stream, but that would require materializing the streams here, which I want to avoid doing prematurely.